File: Ch2/Java/DeleteMiddleNode.java
public class DeleteMiddleNode {

    public static void deleteNode(SinglyLinkedList.Node<Character> toRemove) {
        while (toRemove.next.next != null) {
            toRemove.data = toRemove.next.data;
            toRemove = toRemove.next;
        }
        toRemove.data = toRemove.next.data;
        toRemove.next = null;
    }

    public static void main(String[] args) {
        SinglyLinkedList<Character> linkedList = new SinglyLinkedList();
        linkedList.addToBack('A');
        linkedList.addToBack('B');
        linkedList.addToBack('C');
        linkedList.addToBack('D');
        linkedList.addToBack('E');
        linkedList.addToBack('F');
        System.out.printf("Original: %s\n", linkedList);
        int i = 4;
        SinglyLinkedList.Node<Character> toRemove = linkedList.getNodeAtIndex(i);
        System.out.printf("After deleting %s: ", toRemove);
        deleteNode(toRemove);
        System.out.printf("%s\n", linkedList);
    }



}

File: Ch2/C++/DeleteMiddleNode.cpp
#include <iostream>

#include "SinglyLinkedList.h"

void deleteNode(Node<char>* toRemove) {
    while (toRemove->next->next != NULL) {
        toRemove->data = toRemove->next->data;
        toRemove = toRemove->next;
    }
    toRemove->data = toRemove->next->data;
    toRemove->next = NULL;
}

int main() {
    SinglyLinkedList<char>* linkedList = new SinglyLinkedList<char>();
    for (int i = 65; i < 72; i++) {
        linkedList->addToBack(i);
    }

    std::cout << "Original: " << *linkedList << std::endl;
    int i = 4;
    Node<char>* toRemove = linkedList->getNodeAtIndex(i);
    std::cout << "After deleting " << *toRemove;
    deleteNode(toRemove);
    std::cout << ": " << *linkedList << std::endl;

    return 0;
}